File: benchmarks/livermore/k14_pic_1d.c
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1001
#define NREPS 10000

int main(void) {
    double grd[1002], dex[1002], xi[1002];
    double ex[1002], ex1[1002], dex1[1002];
    double vx[1002], xx[1002], rx[1002];
    double rh[2050];
    long ix[1002], ir[1002];
    double spacer[40];

    /* Init spacer */
    double fuzz = 1.2345e-3, buzz = 1.0 + fuzz, fizz = 1.1 * fuzz;
    for (long k = 1; k <= 39; k++) {
        buzz = (1.0 - fuzz) * buzz + fuzz;
        fuzz = -fuzz;
        spacer[k] = (buzz - fizz) * 0.1;
    }
    double flx = spacer[27];

    /* GRD: valid grid indices */
    for (long i = 1; i <= N; i++)
        grd[i] = (double)((i-1) % 512) + 1.5;

    /* Init EX using signel */
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long k = 1; k <= N; k++) {
        buzz = (1.0 - fuzz) * buzz + fuzz;
        fuzz = -fuzz;
        ex[k] = (buzz - fizz) * 0.1;
    }

    /* Init DEX using signel */
    fuzz = 1.2345e-3; buzz = 1.0 + fuzz; fizz = 1.1 * fuzz;
    for (long k = 1; k <= N; k++) {
        buzz = (1.0 - fuzz) * buzz + fuzz;
        fuzz = -fuzz;
        dex[k] = (buzz - fizz) * 0.1;
    }

    /* Zero arrays */
    for (long i = 1; i <= N; i++) {
        vx[i] = 0; xx[i] = 0; xi[i] = 0;
        ex1[i] = 0; dex1[i] = 0; rx[i] = 0;
        ix[i] = 0; ir[i] = 0;
    }
    for (long i = 1; i <= 2049; i++) rh[i] = 0;

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 1; rep <= NREPS; rep++) {
        for (long i = 1; i <= 2049; i++) rh[i] = 0.0;

        for (long k = 1; k <= N; k++) {
            vx[k] = 0.0;
            xx[k] = 0.0;
            ix[k] = (long)grd[k];
            xi[k] = (double)ix[k];
            ex1[k] = ex[ix[k]];
            dex1[k] = dex[ix[k]];
        }
        for (long k = 1; k <= N; k++) {
            vx[k] = vx[k] + ex1[k] + (xx[k] - xi[k]) * dex1[k];
            xx[k] = xx[k] + vx[k] + flx;
            ir[k] = (long)xx[k];
            rx[k] = xx[k] - (double)ir[k];
            ir[k] = (ir[k] & 2047) + 1;
            xx[k] = rx[k] + (double)ir[k];
        }
        for (long k = 1; k <= N; k++) {
            rh[ir[k]]   += 1.0 - rx[k];
            rh[ir[k]+1] += rx[k];
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("xx[1] = %.15e\n", xx[1]);
    return 0;
}

File: benchmarks/livermore/k14_pic_1d.f
C     Kernel 14 -- 1-D Particle-in-Cell
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K14
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION VX(1001), XX(1001), XI(1001), EX1(1001), DEX1(1001)
      DIMENSION EX(1001), DEX(1001), GRD(1001), RX(1001), RH(2049)
      INTEGER IX(1001), IR(1001)
      DIMENSION SPACER(39)
C
C     Initialize SPACER using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      SCALED = 0.1D0
      DO 5 K = 1, 39
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SPACER(K) = (BUZZ - FIZZ) * SCALED
    5 CONTINUE
      FLX = SPACER(27)
C
C     GRD must be valid grid indices
C
      DO 10 I = 1, 1001
         GRD(I) = DBLE(MOD(I-1, 512)) + 1.5D0
   10 CONTINUE
C
C     Initialize EX, DEX using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 15 K = 1, 1001
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         EX(K) = (BUZZ - FIZZ) * SCALED
   15 CONTINUE
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 17 K = 1, 1001
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         DEX(K) = (BUZZ - FIZZ) * SCALED
   17 CONTINUE
C
C     Zero out arrays
C
      DO 20 I = 1, 1001
         VX(I)   = 0.0D0
         XX(I)   = 0.0D0
         XI(I)   = 0.0D0
         EX1(I)  = 0.0D0
         DEX1(I) = 0.0D0
         RX(I)   = 0.0D0
         IX(I)   = 0
         IR(I)   = 0
   20 CONTINUE
      DO 25 I = 1, 2049
         RH(I) = 0.0D0
   25 CONTINUE
C
      N = 1001
C
C     Kernel 14: 1-D PIC
C
      DO 500 IREP = 1, 10000
C
C        Reset RH
         DO 110 I = 1, 2049
            RH(I) = 0.0D0
  110    CONTINUE
C
C        First pass: gather field values
         DO 200 K = 1, N
            VX(K)   = 0.0D0
            XX(K)   = 0.0D0
            IX(K)   = INT(GRD(K))
            XI(K)   = DBLE(IX(K))
            EX1(K)  = EX(IX(K))
            DEX1(K) = DEX(IX(K))
  200    CONTINUE
C
C        Second pass: push particles
         DO 300 K = 1, N
            VX(K) = VX(K) + EX1(K) + (XX(K)-XI(K))*DEX1(K)
            XX(K) = XX(K) + VX(K) + FLX
            IR(K) = INT(XX(K))
            RX(K) = XX(K) - DBLE(IR(K))
            IR(K) = IAND(IR(K), 2047) + 1
            XX(K) = RX(K) + DBLE(IR(K))
  300    CONTINUE
C
C        Third pass: deposit charge
         DO 400 K = 1, N
            RH(IR(K))   = RH(IR(K))   + 1.0D0 - RX(K)
            RH(IR(K)+1) = RH(IR(K)+1) + RX(K)
  400    CONTINUE
  500 CONTINUE
C
      WRITE(*,900) XX(1)
  900 FORMAT('K14 1-D PIC: xx(1) = ', E23.15E3)
      END
